Simplify remote storage interface

SomeoneToIgnore opened the pull request:
This is a preparation PR that straightens out basic remote storage declarations and adjusts pageserver and safekeeper code to work with the new types.

* makes the "remote path" represented as a relative `PathBuf`, which allows to calculate such remote paths from local paths easily
* removes all "remote path" calculation methods from the `RemoteStorage` trait
* use more explicit types in `IndexPart` to show that layer file names are stored there
* removes unused `missing_layers` field from `IndexPart` strict and keeps the tests that deserialize the old json with that field, proving that it's ok to do so
* replaces `Arc<dyn RemoteStorage>` with a static dispatch, as a preparation for testing `#![feature(async_fn_in_trait)]` nightly feature: it cannot work with `dyn` objects yet. To be able to test the feature fully, `type Target = dyn RemoteStorage;` has to be removed, that will be attempted later.
This also allow to clone underlying `LocalFs` (and `Noop` variant later) without `Arc`.

problame reviewed (changes requested):
Nice cleanups!

---

Review commit by commit:

`Replace dynamic dispatch` ... :+1: can merge as is, don't squash.

`Remove 'local_path' conversion method` :+1: , can merge as is, don't squash. Commit message should make it clear in first sentence that this function is unused. Why didn't we get compiler warnings about this?

`Remove RelativePath conversions`:
-  `RelativePath::new` should use proper assert, not debug assert.
- `strip_base_path` with two arguments with same type seems like a footgun. Can't we turn this function into an `Ext` trait for `&Path`? Like
  ```
  trait IntoRelativePath { fn into_relative_path(&self, base: &Path) }
  Impl<'a> IntoRelativePath for &'a Path {
     ... whatever we have in strip_base_path right now ...
  }
  ````

`Rework remote storage interface`: not reviewed yet.

`Port PS and SK ...`: not reviewed yet.

`Use file names in IndexPart`: not reviewed yet.

`Better naming and docs`: +1, don't squash.

`Remove missing fields` :-1:  see https://github.com/neondatabase/neon/pull/2985 , not as straight-forward I'm afraid.

---

Please re-order the changes that are :+1: and create a separate PR for them. They're a lot of distracting churn that I'd like to get rid of before reviewing the commits that I maked as `not reviewed yet`.

SomeoneToIgnore commented:
> :+1: 

Created https://github.com/neondatabase/neon/pull/2997 and rebased on top, dropped the :-1: one during rebase, thanks for the link.
Excluded better naming and docs, since it's added on top of `RemotePath` which is not yet present in the extracted commits.

problame reviewed (changes requested):
I reviewed the PR as a squashed changed now that the preliminary refactorings have been merged into `main`.

Apart from the inline comments, here's my understanding of what's changed (please correct me and then include it in the squashed commit message):

---

Subject: rework remote_storage interface

Changes:

- Remove `RemoteObjectId` concept from `remote_storage`.
  Operate directly on `/`-separated names instead.
  These names are now represented by `struct RemotePath` which was renamed from `struct RelativePath`.

- Centralize conversion between local & remote paths in `PageServerConf`.
  Previously, the `remote_storage` implementations would do the conversion (`RemoteObjectId`),
  and therefore needed to be initialized with a `working_directory` so that they know what prefix to strip. For example, in pageserver, `working_directory` is the abspath to the local repository dir which contains the `tenants` directory.

  - This is a little infectuous to the pageserver's `storage_sync` interface since we now need to funnel through the `PageServerConf`.

---

Did I miss anything?

SomeoneToIgnore commented on pageserver/src/config.rs:
```diff
@@ -456,6 +456,28 @@ impl PageServerConf {
+    pub fn remote_layer_path(&self, local_layer_path: &Path) -> anyhow::Result<RemotePath> {
+        local_layer_path
+            .strip_prefix(&self.workdir)
+            .context("Failed to strip workdir prefix")
```
It does, in the 2nd context place below (`with_context`) that is more general:

```rust
use anyhow::Context;

fn main() {
    let convert_to_u64 = |int| -> anyhow::Result<u64> {
        u64::try_from(int)
            .context("1st error context")
            .and_then(|_| Err(anyhow::anyhow!("2nd error")))
            .with_context(|| "general_context".to_string())
    };

    println!("{:?}", convert_to_u64(-2));
    println!("=======");
    println!("{:?}", convert_to_u64(2));
}
```

outputs

```
Err(general_context

Caused by:
    0: 1st error context
    1: out of range integral type conversion attempted)
=======
Err(general_context

Caused by:
    2nd error)
```

problame replied:
Sure. What I'm meaning to say is that we never **fail** to strip the workdir prefix. Either the prefix is there (ok) or it isn't (caller's fault, they're only supposed to call us with a path that has the prefix).

So, abetter error would be

```
anyhow::bail!("not a local path")
```

But it's complete hairsplitting at this point :)

SomeoneToIgnore commented on pageserver/src/storage_sync2/index.rs:
```diff
@@ -135,14 +105,19 @@ impl IndexPart {
+                    timeline_layers.insert(layer_name.to_owned());
+                    layer_metadata.insert(layer_name.to_owned(), metadata);
+                }
+                None => error!("Layer {remote_path:?} has no file name, skipping"),
```
Went on with a panic approach with the TODO comment to add more typization there.

SomeoneToIgnore commented:
I think, the general understanding is right, one more bullet to add:
* `struct RelativePath` was moved from `storage_sync2/index.rs` away to remote storage crate, index now holds `String` to (weakly) emphasize this is a layer name

I would argue with the formulation of the 2nd bullet: instead of centralizing everything in `PageServerConf`, I've aimed to represent the remote storage objects with relative paths, which we do anyway, implicitly, converting local paths into remove and back with just using the prefixes.
Safekeeper uses inplace conversions and I think it's less convenient, but I see no issues with `&'static` object. In the end, we can clone the workdir path and use it in all related manipulations.

problame reviewed (approved):
LGTM, the `RemoteObjectId` doesn't make sense for our use case.

I'd prefer if one more person looked at this though, as I don't have the background on why we had `RemoveObjectId` in the first place.

NB: A good portion of the negative diff comes from removing the (pretty dull) unit test in `local_fs.rs` and `s3_bucket.rs`, though. Not sure what they were supposed to cover.

petuhovskiy commented on pageserver/src/config.rs:
```diff
@@ -456,6 +456,28 @@ impl PageServerConf {
             .join(METADATA_FILE_NAME)
     }
 
+    /// Files on the remote stoage are stored with paths, relative to the workdir.
```
```suggestion
    /// Files on the remote storage are stored with paths, relative to the workdir.
```

LizardWizzard commented on libs/remote_storage/src/s3_bucket.rs:
```diff
@@ -408,38 +388,38 @@ impl RemoteStorage for S3Bucket {
     }
 
-    async fn delete(&self, remote_object_id: &RemoteObjectId) -> anyhow::Result<()> {
+    async fn delete(&self, remote_object_id: &RemotePath) -> anyhow::Result<()> {
```
NIT: Probably argument name should be just `path`

LizardWizzard commented on pageserver/src/storage_sync2.rs:
```diff
@@ -366,14 +370,19 @@ impl UploadQueue {
-        for path in &index_part.timeline_layers {
+        let mut files = HashMap::with_capacity(index_part.timeline_layers.len());
+        let timeline_path = conf.timeline_path(&timeline_id, &tenant_id);
+        for timeline_name in &index_part.timeline_layers {
```
Looks like something is wrong with naming. Should it be layer_file_name instead of timeline_name?

LizardWizzard replied:
IndexPart should already contain remote paths, why do we need to do some conversions here?

SomeoneToIgnore replied:
> IndexPart should already contain remote paths, why do we need to do some conversions here?

Not anymore, they've got removed in this PR actually, to emphasize that we list only layer (file) names in the IndexPart.

LizardWizzard reviewed (commented):
Overall looks good to me, mostly minor comments. 

I agree that we should have structs for paths. I had a gnarly bug when I used relative path instead of an absolute one and passed it to remote storage method. And the test contained a race, the bug reproduced when we were able to upload a file during pageserver init.

Also not sure that pageserver config is the right place for .local_path/.remote_path helpers

SomeoneToIgnore commented on libs/remote_storage/src/lib.rs:
```diff
@@ -41,44 +41,27 @@ pub const DEFAULT_REMOTE_STORAGE_S3_CONCURRENCY_LIMIT: usize = 100;
+            relative_path.is_relative(),
+            "Path {relative_path:?} is not relative"
+        );
+        Ok(Self(relative_path.to_path_buf()))
```
I think we could do both (I generally agree with the idea to avoid internal clones), but in this particular case, **all** places actually use `&Path` since produce such relative path via `strip_prefix`.
So I'll leave things as is for now, at least.

SomeoneToIgnore commented on libs/remote_storage/src/lib.rs:
```diff
@@ -433,21 +365,18 @@ mod tests {
-        let k = RemoteObjectId("".to_owned());
-        assert_eq!(k.object_name(), None);
-
-        let k = RemoteObjectId("/".to_owned());
```
Thank you.
I'll add the assert, but won't touch the errors just yet, since https://github.com/neondatabase/neon/pull/3026 is here and presumably we will conflict even more when I start doing the error refactoring.
